subgraph-js/src/types.ts

```diff
@@ -62,6 +69,99 @@ function addTypeNameToPossibleReturn<T>(
   return maybeObject as null | (T & { __typename: string });
 }
 
+function maybeAddTypeNameToPossibleReturn<T>(
+  maybeObject: PromiseOrValue<null | T>,
+  typename: string,
+): PromiseOrValue<null | (T & { __typename?: string })> {
+  if (isPromise(maybeObject)) {
+    return maybeObject.then((x: any) => addTypeNameToPossibleReturn(x, typename));
+  }
+  return addTypeNameToPossibleReturn(maybeObject, typename);
+}
+
+/**
+ * Copied from GraphQL-js.
+ *
+ * For @key on interfaces, we need to check that we can resolve the runtime type of the object returned by the interface
+ * `__resolveReference`. If we cannot, and we simply don't add any `__typename` to the result, the graphQL-js will end
+ * erroring out, but the error will not be user friendly as it will say something along the lines of:
+ * ```
+ * Abstract type "_Entity" must resolve to an Object type at runtime for field "Query._entities". Either the "_Entity" type
+ * should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.
+ * ```
+ * But this is ultimately incorrect, as it is only interface type the user must use "resolveType", add a __typename, or rely
+ * on "isTypeOf". And so we have to somewhat copy this. As we do, we also adapt the message a bit to be more user friendly.
+ */
+function ensureValidRuntimeType(
```

Comment:
> Assuming that this is copied because graphql-js doesn't export it?

It isn't exported, so it certainly forces our hand. But this isn't a direct copy, it's also adapted, mostly to adapt the error message so that actually point to the problem, but there is also some adaptation to the arguments (due to us not having the full execution context that the original method use).

> Can we maybe put this in its own file so that if that changes in the future it's easier to identify?

As said, there is enough modifications to the method in practice that if anyone wanted to compare them later, it's have to be a manual check and I don't think having it in a separate file help. Nor do I think it's worth devoting time to trying to track future changes in the first place.